### PR TITLE
llvm_asm -> asm: arty-e21

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

Convert to asm macro.

#1923

### Testing Strategy

Running apps on arty-e21.

### TODO or Help Wanted

n/a

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by hudson-ayers — COMMENTED** [review]
Nice to get rid of  `cfg[]` + assembly in a couple places!

File: arch/rv32i/src/support.rs [hunk 1]
```diff
@@ -8,15 +8,15 @@ use core::ops::FnOnce;
...
+        asm!("nop", options(nomem, nostack));
     }
 }
 
 #[cfg(all(target_arch = "riscv32", target_os = "none"))]
 #[inline(always)]
 /// WFI instruction
 pub unsafe fn wfi() {
-    llvm_asm!("wfi" :::: "volatile");
+    asm!("wfi", options(nomem, nostack));
```
[the remark this answers is not in the document]
Comment (lschuermann) [reply arch/rv32i/src/support.rs:19]: I don't think so. When we're catching an interrupt in kernel mode and calling the trap handler, which do push some caller-saved registers to the stack. Furthermore, the actual interrupt handler itself might further use the stack. However, prior to continuing execution at the previous location (saved in `mepc`), we're restoring the stack to its original state. So from the point of view of this instruction, the stack won't be modified.

Furthermore, an interrupt can come in at virtual any instruction. A `wfi` is just an explicit way to allow the processor to go into a low-power mode until an interrupt occurs and thereby doesn't behave differently with respect to side-effects compared to a `nop` (which, if I'm not mistaken, is actually a legal implementation of `wfi`).

File: chips/arty_e21_chip/src/chip.rs [hunk 2]
```diff
@@ -134,11 +121,8 @@ impl<'a, I: InterruptService<()> + 'a> ArtyExx<'a, I> {
             addi t0, t0, %lo(_start_trap)
```
Comment (lschuermann) [inline chips/arty_e21_chip/src/chip.rs:121]: I think a `li t0, _start_trap`, which will expand to a `lui` and `addi`, is a bit nicer.

File: chips/arty_e21_chip/src/chip.rs [hunk 3]
```diff
@@ -134,11 +121,8 @@ impl<'a, I: InterruptService<()> + 'a> ArtyExx<'a, I> {
...
             ori  t0, t0, 0x02 // Set CLIC direct mode
             csrw 0x305, t0    // Write the mtvec CSR.
-        "
-        :
-        :
-        :
-        : "volatile");
+            "
```
Comment (lschuermann) [inline chips/arty_e21_chip/src/chip.rs:124]: This is more a general question: shouldn't we mark the register `t0` as clobbered here (as described [here](https://doc.rust-lang.org/beta/unstable-book/library-features/asm.html#clobbered-registers))?

Given that this entire function is a large assembly block, it's probably fine, but I imagine the compiler could potentially do some optimizations or perform inlining which could break the assumptions made about registers not being clobbered in the future?

File: arch/rv32i/src/support.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (hudson-ayers) [reply arch/rv32i/src/support.rs:19]: Hmm, I think you are right. For cortex-m we did not mark `wfi` as `nostack` -- but we did mark it as `preserves_flags`. We should probably make them match. -- I think that both `nostack` and `preserves_flags` both seem right to include.

File: chips/arty_e21_chip/src/chip.rs (hunk 3 above)
Comment (hudson-ayers) [reply chips/arty_e21_chip/src/chip.rs:124]: I agree, t0 should be marked as clobbered.

**Review by hudson-ayers — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Review by alistair23 — APPROVED** [review]

**Pull request merged**